File: 545. Boundary of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def boundaryOfBinaryTree(self, root: TreeNode) -> List[int]:
        '''
        Time O(n)
        Space O(n) if we count all list to return
        '''
        if not root:
            return []
        self.res = []
        self.rig = []
        self.le = []
        self.findleftmost(root.left)
        self.findrightmost(root.right)
        self.findleaf(root.left)
        self.findleaf(root.right)
        
        return [root.val] + self.res + self.le + self.rig[::-1]
        
    def findleftmost(self, node):
        if not node or (not node.left and not node.right):
            return
        self.res.append(node.val)
        if node.left:
            self.findleftmost(node.left)
        else:
            self.findleftmost(node.right)
            
    def findrightmost(self, node):
        if not node or (not node.left and not node.right):
            return
        self.rig.append(node.val)
        if node.right:
            self.findrightmost(node.right)
        else:
            self.findrightmost(node.left)
            
    def findleaf(self, node):
        if not node:
            return
        if not node.left and not node.right:
            self.le.append(node.val)
            return 
        if node.left:
            self.findleaf(node.left)
        if node.right:
            self.findleaf(node.right)

File: 545. Boundary of Binary Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
/**
 * O(n) time O(n) space
 */
class Solution {
    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        List<Integer> ret = new ArrayList<>();
        if (root == null) {
            return ret;
        }
        //consider about duplication of root, root may exist in left, right and leaves
        ret.add(root.val);
        leftBoundary(root.left, ret);
        leaves(root.left, ret);
        leaves(root.right, ret);
        rightBoundary(root.right, ret);
        return ret;
    }
    private void leftBoundary(TreeNode root, List<Integer> ret) {
        if (root == null || (root.left == null && root.right == null)) {
            return;
        }
        ret.add(root.val);
        if (root.left != null) {
            leftBoundary(root.left, ret);
        } else {
            leftBoundary(root.right, ret);
        }
    }
    
    private void rightBoundary(TreeNode root, List<Integer> ret) {
        if (root == null || (root.left == null && root.right == null)) {
            return;
        }
        if (root.right != null) {
            rightBoundary(root.right, ret);
        } else {
            rightBoundary(root.left, ret);
        }
        ret.add(root.val);
    }
    
    
    private void leaves(TreeNode root, List<Integer> ret) {
        if (root == null) {
            return;
        }
        if (root.left == null && root.right == null) {
            ret.add(root.val);
        } else {
            if (root.left != null) {
                leaves(root.left, ret);
            }
            if (root.right != null) {
                leaves(root.right, ret);
            }
        }
    }
}